backend-lib/src/main/rust/lib.rs

```diff
@@ -1148,23 +1016,245 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_re
+            (0..count)
+                .map(|i| {
+                    env.get_object_array_element(roots, i)
+                        .map_err(|e| e.into())
+                        .and_then(|jobj| decode_sapling_subtree_root(&env, jobj))
+                })
+                .collect::<Result<Vec<_>, _>>()?
+        };
+
+        db_data
+            .put_sapling_subtree_roots(start_index, &roots)
+            .map(|()| JNI_TRUE)
+            .map_err(|e| format_err!("Error while storing Sapling subtree roots: {}", e))
+    });
+
+    unwrap_exc_or(&env, res, JNI_FALSE)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_updateChainTip(
+    env: JNIEnv<'_>,
+    _: JClass<'_>,
+    db_data: JString<'_>,
+    height: jlong,
+    network_id: jint,
+) -> jboolean {
+    let res = panic::catch_unwind(|| {
+        let network = parse_network(network_id as u32)?;
+        let mut db_data = wallet_db(&env, network, db_data)?;
+        let height = BlockHeight::try_from(height)?;
+
+        db_data
+            .update_chain_tip(height)
+            .map(|()| JNI_TRUE)
+            .map_err(|e| format_err!("Error while updating chain tip to height {}: {}", height, e))
+    });
+
+    unwrap_exc_or(&env, res, JNI_FALSE)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_getFullyScannedHeight(
+    env: JNIEnv<'_>,
+    _: JClass<'_>,
+    db_data: JString<'_>,
+    network_id: jint,
+) -> jlong {
+    let res = panic::catch_unwind(|| {
+        let network = parse_network(network_id as u32)?;
+        let db_data = wallet_db(&env, network, db_data)?;
+
+        match db_data.block_fully_scanned() {
+            Ok(Some(metadata)) => Ok(i64::from(u32::from(metadata.block_height()))),
+            // Use -1 to return null across the FFI.
+            Ok(None) => Ok(-1),
+            Err(e) => Err(format_err!(
+                "Failed to read block metadata from WalletDb: {:?}",
+                e
+            )),
+        }
+    });
+    unwrap_exc_or(&env, res, -1)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_getMaxScannedHeight(
+    env: JNIEnv<'_>,
+    _: JClass<'_>,
+    db_data: JString<'_>,
+    network_id: jint,
+) -> jlong {
+    let res = panic::catch_unwind(|| {
+        let network = parse_network(network_id as u32)?;
+        let db_data = wallet_db(&env, network, db_data)?;
+
+        match db_data.block_max_scanned() {
+            Ok(Some(metadata)) => Ok(i64::from(u32::from(metadata.block_height()))),
+            // Use -1 to return null across the FFI.
+            Ok(None) => Ok(-1),
+            Err(e) => Err(format_err!(
+                "Failed to read block metadata from WalletDb: {:?}",
+                e
+            )),
+        }
+    });
+    unwrap_exc_or(&env, res, -1)
+}
+
+fn encode_scan_progress(env: &JNIEnv<'_>, progress: Ratio<u64>) -> Result<jobject, failure::Error> {
```

Comment:
Since the Rust type [does not enforce](https://github.com/zcash/librustzcash/blob/1e39daf6f83c4b62d645e74385d24299dc4414c7/zcash_client_backend/src/data_api.rs#L131-L136) any constraints on the numerator and denominator, the constraints imposed by [JniScanProgress](https://github.com/zcash/zcash-android-wallet-sdk/pull/1116/files#diff-7762a27caf428321f053453236c0a045b09f17f84385cf953891b989bb714656R17-R28) should be documented (even though this function is internal).
```suggestion
/// Returns a `JniScanProgress` object, provided that numerator is nonnegative, denominator
/// is positive, and the represented ratio is in the range 0.0 to 1.0 inclusive.
fn encode_scan_progress(env: &JNIEnv<'_>, progress: Ratio<u64>) -> Result<jobject, failure::Error> {
```